### ISSUE TITLE
Unexpected deadlock when starting a MutableProperty's SignalProducer

### ISSUE THREAD

**Issue by ianthehenry** [ReactiveCocoa/ReactiveCocoa#2562]
```swift
let foo: MutableProperty<Int> = MutableProperty(0)
let bar: MutableProperty<Int> = MutableProperty(0)

foo.producer.skip(1).startWithNext { _ in
    bar.value = 0
}

bar.producer.skip(1).startWithNext { _ in
    foo.producer.startWithNext { _ in
        print("whoo")
    }
}

foo.value = 0
```

In `4.0.4-alpha-1`, this deadlocks, but I would expect it to print `"whoo"`. The following similar-looking constructs produce no deadlock:

```swift
foo.producer.skip(1).startWithNext { _ in
    foo.producer.startWithNext { _ in
        print("whoo")
    }
}
```

```swift
bar.producer.skip(1).startWithNext { _ in
    print(foo.value)
}
```

This is a very minimized version of something that happened in application code where accessing a `lazy` property caused the subscribe-within-the-subscribe.

**Comment by andersio** [comment]
This is a known limitation (#2515) of the current implementation of `SignalProducer.buffer`.

**Comment by andersio** [comment]
After revisiting my previous answer, I think an elaboration might be needed to avoid confusions, and I apologise in advance since I missed a few critical lines.

As mentioned before, starting a first buffered producer as a result of an observed event (e.g. `Next`) from a second producer is impossible at this moment. This is not true, unless you are in a specific circumstance, that is the event is sent synchronously as a result of an observed event from the *first producer*.

Specifically, the deadlock is caused by things happened behind `foo.value = 0`. Some kind of **indirect** recursion happened there, and with the current implementation it would always deadlock.

Why? It is because the *buffering observer* and the *start handler* of the buffering producers have `dispatch_sync` calls to its internal queue, synchronising works behind the scene. It is designed to elide only recursion on itself, but not across multiple ones in this case. i.e. the call stack in a reduced form:

```
// foo.value = 0 is called.
// The setter sends the new value to `foo.producer`'s buffering observer, which queues works to its internal queue and waits for completion.
dispatch_sync(foo_internal_queue) {
    // Foo calls its observer, and its observer sets `bar.value` to `0`.
    // The setter sends the new value to the corresponding buffering observer, which queues works to its internal queue and waits for completion.
    dispatch_sync(bar_internal_queue) {
        // Bar calls its observer, and its observer starts again buffering producer Foo.
        // Foo's start handler queues works to its internal queue and waits for completion.
        dispatch_sync(foo_internal_queue) {
             // This block would never be executed, since the outmost and this dispatch_sync call would be waiting on each other. i.e. deadlock.
        }
    }
}
```

As said in #2515, the only way is breaking it with asynchrony manually (`observeOn`, `startOn`), or avoid it in the first place. In RAC's design guideline, synchrony is the guaranteed default behaviour. So I would say this is an expected limitation, and likely hard to "fix". (queue drain was proposed but written off)

**Comment by mdiep** [comment]
Closing because this is a duplicate of #2515, which was fixed in #2900.
